### PR TITLE
release v4.4.0

### PR DESCRIPTION
**Pull request by jameslamb**
### Release checklist:

Copied from #6277, with a few changes.

*before merge*

- [x] Update [VERSION.txt](https://github.com/microsoft/LightGBM/blob/master/VERSION.txt) number.
- [x] Update version in [Appveyor config file](https://github.com/microsoft/LightGBM/blob/master/.appveyor.yml).
- [x] Update version in [`configure` file of R-package](https://github.com/microsoft/LightGBM/blob/master/R-package/configure): `/gha run r-configure`.
- [x] Change `development.mode` from `unreleased` to `release` in [pkgdown config file](https://github.com/microsoft/LightGBM/blob/master/R-package/pkgdown/_pkgdown.yml).
- [x] Update version in `python-package/pyproject.toml`
- [x] Add release branch to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/), trigger a new build, check docs
- [x] All docs for new behavior have Sphinx `versionadded::` annotations ([docs on those](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded))
- [x] All new parameters in `config.h` have `*New in version {version}*` comments added
- [x] Run the valgrind checks with `/gha run r-valgrind` ([docs link](https://github.com/microsoft/LightGBM/blob/master/R-package/README.md#valgrind)) 
- [x] test Python and R packages on arm64 Mac

Items to be addressed before merging:

* [x] #6432
* [x] #6409
* [x] #6402
* [x] #6274
* [x] #6454
  - [x] #6467
* [x] #6377
* [x] #6442
* deprecation warnings for these:
  - [x] #6435
  - [x] #6436

I'd really like for this to be the first release with `arm64` macOS wheels, if we can get there. But let's not delay the release waiting on that if everything else above is done.

* [x] #6391

*after merge*

- [ ] Push `v*` tag to trigger `GitHubRelease` action at Azure Pipelines.
- [ ] Update [`stable` tag at GitHub](https://github.com/microsoft/LightGBM/releases/tag/stable).

<details><summary>code to update tags (click me)</summary>

```shell
NEW_VERSION="4.4.0"
git fetch upstream --tags
git tag -d stable
git push upstream :refs/tags/stable
git tag stable
git tag "v${NEW_VERSION}"
git push upstream stable "v${NEW_VERSION}"
```

</details>

- [ ] Add release notes from `release-drafter` to the automatically-created GitHub release, convert it from `Draft` to published.
- [ ] Add new tag to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/) and trigger a new build.
- [ ] Upload release to [CRAN](https://cran.r-project.org/package=lightgbm)
- [ ] Remove the release branch from RTD versions
- [ ] Upload release to [test PyPI](https://test.pypi.org/project/lightgbm/)
- [ ] Upload release to [PyPI](https://pypi.org/project/lightgbm/).

<details><summary>code to publish to PyPI (summary)</summary>

```shell
mkdir /tmp/lgb-release
cd /tmp/lgb-release

NEW_VERSION="4.4.0"

# NOTE: requires a GitHub personal access token with "repo" scope
gh release download \
    --repo microsoft/LightGBM \
    --dir ./artifacts \
    --pattern 'lightgbm*-py3-*.whl' \
    --pattern "lightgbm-${NEW_VERSION}.tar.gz" \
    "v${NEW_VERSION}"

twine upload \
    -r testpypi \
    ./artifacts/*
```

*(`gh` is the GitHub CLI, see https://cli.github.com/manual/gh_release_download)*

Then confirmed that installing the latest wheel works.

```shell
pip install -i https://test.pypi.org/simple/ "lightgbm==${NEW_VERSION}"
python ./examples/python-guide/logistic_regression.py
```

Then pushed them to real PyPI.

```shell
twine upload \
    ./artifacts/*
```

</details>

- [ ] Update [cran-comments.md](https://github.com/microsoft/LightGBM/blob/master/R-package/cran-comments.md) when new release is accepted on CRAN.
- [ ] Upload release to [NuGet](https://www.nuget.org/packages/LightGBM/).
- [ ] Update version and commit hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/l/lightgbm.rb) ([docs](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md#to-submit-a-version-upgrade-for-the-foo-formula))
- [ ] put up another PR incrementing development version here in GitHub (example: #6090)

<details><summary>code to open that dev version PR (click me)</summary>

```shell
RELEASE_PR_NUMBER="6439"
RELEASE_VERSION=$(cat ./VERSION.txt)
DEV_VERSION="${RELEASE_VERSION}.99"

git checkout -b ci/dev-version

echo "${DEV_VERSION}" > ./VERSION.txt
sed \
    -i .bak \
    "s|${RELEASE_VERSION}|${DEV_VERSION}|g" \
    .appveyor.yml

sed \
    -i .bak \
    "s|version = \"${RELEASE_VERSION}\"|version = \"${DEV_VERSION}\"|g" \
    ./python-package/pyproject.toml

sed \
    -i .bak \
    's|mode\: release|mode\: unreleased|g' \
    ./R-package/pkgdown/_pkgdown.yml

docker run \
    --rm \
    -v $(pwd):/opt/LightGBM \
    -w /opt/LightGBM \
    ubuntu:22.04 \
    ./R-package/recreate-configure.sh

git add \
    ./.appveyor.yml \
    ./R-package/configure \
    ./R-package/pkgdown/_pkgdown.yml \
    ./VERSION.txt \
    ./python-package/pyproject.toml

commit_msg="bump development version to ${DEV_VERSION}"
git commit -m "${commit_msg}"
git push upstream ci/dev-version

gh pr create \
    --repo microsoft/LightGBM \
    --base 'master' \
    --label 'maintenance' \
    --title "${commit_msg}" \
    --body "now that v${RELEASE_VERSION} has been released: #${RELEASE_PR_NUMBER}"
```

</details>

### Notes for Reviewers

I believe this should be `v4.4.0` instead of `v4.3.1` because of some small breaking changes:

* #6418
* #6327
* #6293

See https://github.com/microsoft/LightGBM/releases.

### CODE REVIEW

**Comment by jameslamb** [comment]
@shiyu1994 @guolinke @jmoralez @borchero @btrotta to help with the release, could you please do the following?

* comment here if you want to change anything about the proposed list of issues/PRs to be addressed before merging
* comment on the open Request for Comment issues:
  - #6435
  - #6436
* share any other thoughts here about this release (e.g. the version number, the timing, the testing plan)

Thank you!

**Comment by borchero** [comment]
Plan sounds great to me, thanks a lot for taking care of the release @jameslamb!

I think that `v4.4.0` makes sense as the next version since we're introducing a few new features :)

**Comment by shiyu1994** [comment]
Thank you @jameslamb for the releasing. I don't think we have anything urgent to be included in v4.4.0 for now.

**Comment by celestinoxp** [comment]
Please add verbose fix in this new version...

**Comment by jameslamb** [comment]
I've added #6454 to the list of issues to resolve for this release. I'll prioritize that this week.

Added that because `numpy` 2.0 will be released June 16th: https://github.com/numpy/numpy/issues/26191#issuecomment-2125646061. I think we should try to have a `lightgbm` release out by then that's compatible.

**Comment by jameslamb** [comment]
Alright I think we are close!

I'm sorry for the rush, but could we please try to finish the release in the next 3 days, so we could hopefully put up a release of the Python package prior to `numpy` 2.0 going out on June 16th?

I think we can get it done.

I can do all of the release tasks to get this PR ready, but could use some help on reviews of the following:

* https://github.com/microsoft/LightGBM/pull/6467
* https://github.com/microsoft/LightGBM/pull/6391

@guolinke @shiyu1994 @borchero @jmoralez if you have time in the next few days, could you help with reviews on those?

While that's going on, I'll continue trying to help @jmoralez unblock R CI in #6442, and I can review that PR and #6377.

**Comment by jameslamb** [comment]
> All docs for new behavior have Sphinx `versionadded::` annotations
> All new parameters in config.h have `*New in version {version}*` comments added

I went through the draft release notes at https://github.com/microsoft/LightGBM/releases to look for PRs in this release that added new functionality. I believe only these need docs notes saying "added in v4.4.0":

* #6398
* #6293
* #6274

Pushed those in https://github.com/microsoft/LightGBM/pull/6439/commits/ebd8c984c88b734c1947cbcdfadf5deaf3356f96

> Add release branch to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/), trigger a new build, check docs

* build: https://readthedocs.org/projects/lightgbm/builds/24691521/
* docs:
   - main: https://lightgbm.readthedocs.io/en/release-v4.4.0/
   - R: https://lightgbm.readthedocs.io/en/release-v4.4.0/R/reference/

Here's an example of the new version-added notes rendering: https://lightgbm.readthedocs.io/en/release-v4.4.0/R/reference/lgb.save.html

Screenshot 2024-06-13 at 11 48 24 PM

**Comment by jameslamb** [comment]
> test Python and R packages on M1/M2/M3 Mac

LightGBM now has a CI job that does this on every commit (as of #6391).

I tested the R package on my M2 Mac tonight, and put up an issue documenting the need to add a CI job in the future: #6481

**Comment by jameslamb** [comment]
Alright I think this is ready!

@guolinke @shiyu1994 @jmoralez @borchero could you please take a look when you have time?

Once at least 2 people have approved (and at least one of @guolinke and @shiyu1994 ), I'll merge this and proceed with the release (pushing tags, publishing release notes, etc.).

**Review by borchero — APPROVED** [review]
Nice, thanks for getting this over the line @jameslamb! Sorry for being inactive the last few days... 😬

**Review by guolinke — APPROVED** [review]

File: include/LightGBM/config.h
Comment (jmoralez) [inline include/LightGBM/config.h:395]: This should be below (on early_stopping_min_delta)

**Review by jmoralez — CHANGES_REQUESTED** [review]

File: .appveyor.yml
Comment (jameslamb) [inline .appveyor.yml:1]:
```suggestion
version: 4.4.0.{build}
```

oops (this is necessary to keep getting new builds)

File: include/LightGBM/config.h
Comment (jameslamb) [reply include/LightGBM/config.h]: thank you for catching this! I just pushed a commit fixing it.

**Comment by jameslamb** [comment]
> Sorry for being inactive the last few days...

Don't worry about it @borchero ! We're mostly all volunteers here, with availability that varies week-to-week. Appreciate all your help on the other contributions that led up to this release being almost-ready today 😁

**Review by jmoralez — APPROVED** [review]

**Comment by jameslamb** [comment]
Thanks everyone! Whenever CI passes here, I'll merge this and continue with the release.

**Comment by jameslamb** [comment]
Ran the following to create the `v4.3.0` tag and update the `stable` tag.

```shell
git checkout master
git pull upstream master
git fetch upstream --tags
git tag -d stable
git push upstream :refs/tags/stable
git tag stable
git tag v4.4.0
git push upstream stable v4.4.0
```

*(NOTE: I alias this repo to `upstream` and my fork to `origin` in my `git` settings)*

https://github.com/microsoft/LightGBM/tags

image

That triggered an Azure DevOps build which should create the release automatically: https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16384&view=results

This takes around 90 minutes (because of the QEMU CI job).

I'll check back in a few hours and continue the release then.

**Comment by jameslamb** [comment]
> Add new tag to [RTD versions](https://readthedocs.org/projects/lightgbm/versions/) and trigger a new build.
> Remove the release branch release/v{n}.{n}.{n} from RTD versions

These are done.

* build: https://readthedocs.org/projects/lightgbm/builds/24700459/
* docs: https://lightgbm.readthedocs.io/en/v4.4.0/Parameters.html

And v4.4.0 in the selector:

image

**Comment by jameslamb** [comment]
>  Add release notes from release-drafter to the automatically-created GitHub release, convert it from Draft to published.

The release is now published here on GitHub:

https://github.com/microsoft/LightGBM/releases/tag/v4.4.0

**Comment by jameslamb** [comment]
Submitted to CRAN:

image

**Comment by jameslamb** [comment]
Uploaded to test PyPI and tested installing from it (including running lightgbm's tests) on my M2 Mac. That worked!

So published to real PyPI.

**`v4.4.0` is now up on PyPI**

**Comment by jameslamb** [comment]
put up dev version PR: https://github.com/microsoft/LightGBM/pull/6485

**Comment by jameslamb** [comment]
> upload release to NuGet

done: https://www.nuget.org/packages/LightGBM

**Comment by jameslamb** [comment]
The R package has passed all checks at https://cran.r-project.org/web/checks/check_results_lightgbm.html

image

**Comment by jameslamb** [comment]
> Update [cran-comments.md](https://github.com/microsoft/LightGBM/blob/master/R-package/cran-comments.md) when new release is accepted on CRAN.

Done in #6499

> Update version and commit hash in [Homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/l/lightgbm.rb) ([docs](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md#to-submit-a-version-upgrade-for-the-foo-formula))

This was done in https://github.com/Homebrew/homebrew-core/pull/174635

---

And with that, this release is complete! Thanks everyone!!

**Pull request merged**
